### PR TITLE
Avoid FD and memory leaks

### PR DESCRIPTION
**Pull request by DemiMarie**
See individual commit messages for details.

### CODE REVIEW

File: vchan/init.c [hunk 1]
```diff
@@ -29,30 +29,29 @@
...
-    ctrl->xc_handle = xc_interface_open(NULL, NULL, 0);
-    if (!ctrl->xc_handle) {
-        /* error already logged by xc_interface_open */
-        libxenvchan_close(ctrl->xenvchan);
-        free(ctrl);
-        return NULL;
-    }
-    return ctrl;
+    if ((ctrl->xc_handle = xc_interface_open(NULL, NULL, 0)))
+        return ctrl; /* Everything good */
```
Comment (marmarek) [inline vchan/init.c:46]: I'd prefer conditional handling an error, and final successful `return` be outside of `if`. It means one more goto label, but IMO is easier to read.

File: vchan/io.c [hunk 2]
```diff
@@ -98,12 +99,11 @@ int libvchan_wait(libvchan_t *ctrl) {
...
-            fd_set rd_set;
-            struct timeval tv = { 10, 0 };
             int vchan_fd = libxenvchan_fd_for_select(ctrl->xenvchan);
-            FD_ZERO(&rd_set);
-            FD_SET(vchan_fd, &rd_set);
-            switch (select(vchan_fd+1, &rd_set, NULL, NULL, &tv)) {
+            struct pollfd fds[] = {
+                { .fd = vchan_fd, .events = POLLIN | POLLOUT | POLLPRI | POLLRDHUP, .revents = 0 },
+            };
+            switch (poll(fds, 1, 10)) {
```
Comment (marmarek) [inline vchan/io.c:106]:
```suggestion
            switch (poll(fds, 1, 10000)) {
```

File: vchan/io.c [hunk 3]
```diff
@@ -98,12 +99,11 @@ int libvchan_wait(libvchan_t *ctrl) {
...
          */
         while (ret == -2 && libxenvchan_is_open(ctrl->xenvchan) == 2) {
-            fd_set rd_set;
-            struct timeval tv = { 10, 0 };
             int vchan_fd = libxenvchan_fd_for_select(ctrl->xenvchan);
-            FD_ZERO(&rd_set);
-            FD_SET(vchan_fd, &rd_set);
-            switch (select(vchan_fd+1, &rd_set, NULL, NULL, &tv)) {
+            struct pollfd fds[] = {
+                { .fd = vchan_fd, .events = POLLIN | POLLOUT | POLLPRI | POLLRDHUP, .revents = 0 },
```
Comment (marmarek) [inline vchan/io.c:104]: `PULLRDHUP` is not applicable here. `PULLPRI` looks irrelevant for this FD too. `PULLOUT` is actively wrong here.
So, please keep only `PULLIN`.

**Review by marmarek — CHANGES_REQUESTED** [review]

File: vchan/init.c [hunk 4]
```diff
@@ -29,30 +29,29 @@
...
     ctrl->remote_domain = domain;
-    ctrl->xc_handle = xc_interface_open(NULL, NULL, 0);
-    if (!ctrl->xc_handle) {
-        /* error already logged by xc_interface_open */
-        libxenvchan_close(ctrl->xenvchan);
-        free(ctrl);
-        return NULL;
-    }
+    if (!(ctrl->xc_handle = xc_interface_open(NULL, NULL, 0)))
+        goto err_xc_open;
```
Comment (marmarek) [inline vchan/init.c:46]: trailing space

**Review by marmarek — CHANGES_REQUESTED** [review]

**Comment by DemiMarie** [comment]
PipelineRetryFailed

**Review by marmarek — APPROVED** [review]

**Pull request merged**